src/UI/IdentityBuilderUIExtensions.cs

```diff
@@ -51,56 +47,19 @@ private static void AddAdditionalApplicationParts(IdentityBuilder builder)
             // * Discovery of the parts.
             // * Ordering of the parts.
             // * Loading of the assembly in memory.
-            var thisAssembly = typeof(IdentityBuilderUIExtensions).Assembly;
-            var additionalReferences = thisAssembly
-                .GetCustomAttributes<AssemblyMetadataAttribute>()
-                .Where(am => string.Equals(am.Key, "Microsoft.AspNetCore.Mvc.AdditionalReference"))
-                .Select(am => am.Value.Split(',')[0])
-                .ToArray();
 
             var mvcBuilder = builder.Services
                 .AddMvc()
-                .ConfigureApplicationPartManager(apm =>
+                .ConfigureApplicationPartManager(partManager =>
                 {
-                    foreach (var reference in additionalReferences)
+                    var thisAssembly = typeof(IdentityBuilderUIExtensions).Assembly;
+                    var relatedAssemblies = RelatedAssemblyAttribute.GetRelatedAssemblies(thisAssembly, throwOnError: true);
+
+                    foreach (var assembly in relatedAssemblies)
                     {
-                        var fileName = Path.GetFileName(reference);
-                        var filePath = Path.Combine(Path.GetDirectoryName(thisAssembly.Location), fileName);
-                        var additionalAssembly = LoadAssembly(filePath);
-                        // This needs to change to additional assembly part.
-                        var additionalPart = new AssemblyPart(additionalAssembly);
-                        if (!apm.ApplicationParts.Any(ap => HasSameName(ap.Name, additionalPart.Name)))
-                        {
-                            apm.ApplicationParts.Add(additionalPart);
-                        }
+                        partManager.ApplicationParts.Add(new CompiledRazorAssemblyPart(assembly));
```

Comment:
This should be just using the part factory on the related assembly with a different context to light it up 😕. Something along the lines of:
```csharp
var partFactory = ProvidePartFactory.GetPartFactory(assembly);
var parts = partFactory.CreateApplicationParts(assembly,"someCtx");
apm.ApplicationParts.Add(parts)
```

Reply:
Ok. So don't use the `NullPartFactory` and write a custom part factory?

Reply:
Yes.

Orthogonal to this, maybe NullPartFactory should be -> OptionalPartFactory and support this scenario through the ctx parameter.

Reply:
The important bit is that the parts for an assembly always have to be instantiated through the factory in the assembly so that if something changes it only changes in one place.

Reply:
> The important bit is that the parts for an assembly always have to be instantiated through the factory 

Why is this important?

Reply:
Because if not, when something changes in the parts implementation you need to change it in all the consuming clients. For example, if for some reason precompilation changes and starts creating an additional part for something else, then all consuming clients (identity, etc) need to change to create that new part themselves.